### PR TITLE
Fix an issue with smart contract function overloading

### LINKED ISSUE #6837: contract.methods['functionName(address)'](address) is running method with string signature
#### Expected behavior
Expect `contract.methods['functionName(address)'](address).call();` to call the method with the address signature.

#### Actual behavior
Call happens on the method with string signature instead.

#### Steps to reproduce the behavior
```
// in js file
async function getUserData(address) {
    return await userContract.methods['getUserData(address)'](address).call();
}

// in solidity file
function getUserData(address userAddress) public view returns (UserData memory) {
    return _userData[_idOfAddress[userAddress]];
}

function getUserData(string memory userName) public view returns (UserData memory) {
    return  _userData[_idOfName[userName]];
}
```

1. load contract (live on sepolia testnet @ 0x6f380E5d68D7211f305976f86c66a22Ce17C3054)
2. in a browser with the js loaded run "await getUserData('0xF174CC549e39545a44041bbf3Daf94E5b6DD7F54')" in console
3. in a browser with the js loaded run "await getUserData('technicallywebdev')" in console
4. expect first to log userData object, expect 2nd to fail with incorrect type.

#### Logs

output of step 2: 
`{0: 0n, 1: '0x0000000000000000000000000000000000000000', 2: '', 3: '', 4: '', 5: Array(0), __length__: 6, registrationDate: 0n, defaultAddress: '0x0000000000000000000000000000000000000000', userName: '', userBio: '', …}`
output of step 3: 
`{0: 1709018832n, 1: '0xF174CC549e39545a44041bbf3Daf94E5b6DD7F54', 2: 'technicallydev', 3: 'Hello World!', 4: '/assets/images/logo.png', 5: Array(1), __length__: 6, registrationDate: 1709018832n, defaultAddress: '0xF174CC549e39545a44041bbf3Daf94E5b6DD7F54', userName: 'technicallydev', userBio: 'Hello World!', …}`

#### Environment

run client-side via html import. 
npm: 9.8.0, node: 18.6.1 web3 version: https://cdn.jsdelivr.net/npm/web3@latest/dist/web3.min.js
the full project is available on https://github.com/TechnicallyWeb3/technicallyweb3.com/commit/2cddcb69b351271dd680345377000f941329030f

### PR DESCRIPTION
**Pull request by Muhammad-Altabba**
## Description

Please include a summary of the changes and be sure to follow our [Contribution Guidelines](https://github.com/web3/web3.js/blob/4.x/.github/CONTRIBUTING.md).

Fixes #6837

## Type of change

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist:

- [ ] I have selected the correct base branch.
- [ ] I have performed a self-review of my own code.
- [ ] I have commented my code, particularly in hard-to-understand areas.
- [ ] I have made corresponding changes to the documentation.
- [ ] My changes generate no new warnings.
- [ ] Any dependent changes have been merged and published in downstream modules.
- [ ] I ran `npm run lint` with success and extended the tests and types if necessary.
- [ ] I ran `npm run test:unit` with success.
- [ ] I ran `npm run test:coverage` and my test cases cover all the lines and branches of the added code.
- [ ] I ran `npm run build` and tested `dist/web3.min.js` in a browser.
- [ ] I have tested my code on the live network.
- [ ] I have checked the Deploy Preview and it looks correct.
- [ ] I have updated the `CHANGELOG.md` file in the root folder.
- [ ] I have linked Issue(s) with this PR in "Linked Issues" menu.

### CODE REVIEW

File: packages/web3-eth-contract/src/contract.ts [hunk 1]
```diff
@@ -1082,19 +1073,26 @@ export class Contract<Abi extends ContractAbi>
...
+						applicableMethodAbi.push(_abi);
 					} catch (e) {
 						errors.push(e as Web3ValidationErrorObject);
 					}
 				}
+				if (applicableMethodAbi.length === 1) {
+					[methodAbi] = applicableMethodAbi; // take the first item that is the only item in the array
+				} else {
+					[methodAbi] = applicableMethodAbi; // take the first item in the array
+					// TODO: Should throw a new error with the list of methods found.
```
Comment (Muhammad-Altabba) [inline packages/web3-eth-contract/src/contract.ts:1109]: An issue has been created: https://github.com/web3/web3.js/issues/6923
The changes was not applied to not introduce a breaking change. If someone things this is an acceptable change in 4.x. We can discuss and probably implement.

**Review by jdevcs — APPROVED** [review]

File: packages/web3-eth-contract/src/contract.ts [hunk 2]
```diff
@@ -446,11 +446,7 @@ export class Contract<Abi extends ContractAbi>
 	);
 	public constructor(
 		jsonInterface: Abi,
-		addressOrOptionsOrContext?:
-			| Address
-			| ContractInitOptions
-			| Web3ContractContext
-			| Web3Context,
+		addressOrOptionsOrContext?: Address | ContractInitOptions | Web3ContractContext | Web3Context,
```
Comment (jdevcs) [inline packages/web3-eth-contract/src/contract.ts]: will be very helpful to not include such changes , thanks,

**Review by avkos — APPROVED** [review]

File: packages/web3-eth-contract/src/contract.ts (hunk 2 above)
Comment (Muhammad-Altabba) [reply packages/web3-eth-contract/src/contract.ts]: Sorry, those changes came after I saved the file. They are made by prettier according to the current configuration. 
So, either the file was saved early without applying prettier formatting. Or, there is something different on my machine. For now, I will revert those changes back.

**Review by Frank122480 — COMMENTED** [review]
Looks good

**Pull request merged**